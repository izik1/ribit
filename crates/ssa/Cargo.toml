[package]
authors.workspace = true
edition.workspace = true
name = "ribit-ssa"
publish.workspace = true
version.workspace = true

[dependencies]
ribit-core.workspace = true

[dev-dependencies]
expect-test.workspace = true
ribit-asm.workspace = true
