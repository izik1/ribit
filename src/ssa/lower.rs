use std::mem;

use super::{BinOp, CmpKind, Id, Instruction, InstructionId, Source};

use crate::{instruction, opcode, register};

pub struct Context {
    next_id: Id,
    pub pc: Id,
    instructions: Vec<Instruction>,
}

impl Context {
    const fn initial_pc(pc: u32) -> Instruction {
        let instr = Instruction::LoadConst {
            dest: Id(0),
            src: pc,
        };

        instr
    }

    pub fn new(start_pc: u32) -> Self {
        let pc = Self::initial_pc(start_pc);

        Self {
            next_id: Id(1),
            pc: Id(0),
            instructions: vec![pc],
        }
    }

    pub fn add_pc(&mut self, src: Source) -> Id {
        let id = self.new_id();

        self.instructions.push(Instruction::BinOp {
            op: BinOp::Add,
            dest: id,
            src1: src,
            src2: Source::Id(mem::replace(&mut self.pc, id)),
        });

        id
    }

    fn new_id(&mut self) -> Id {
        let id_num = self.next_id.0;
        std::mem::replace(&mut self.next_id, Id(id_num + 1))
    }

    fn instr_with_id<F: FnOnce(Id) -> Instruction>(&mut self, f: F) -> Id {
        let id = self.new_id();
        self.instructions.push(f(id));
        id
    }

    pub fn binop(&mut self, op: BinOp, src1: Source, src2: Source) -> Id {
        self.instr_with_id(|dest| Instruction::BinOp {
            dest,
            src1,
            src2,
            op,
        })
    }

    pub fn read_register(&mut self, reg: register::RiscV) -> Id {
        let id = self.new_id();

        self.instructions
            .push(Instruction::ReadReg { dest: id, src: reg });

        id
    }

    pub fn load_register(&mut self, reg: Option<register::RiscV>) -> Id {
        if let Some(reg) = reg {
            self.read_register(reg)
        } else {
            self.load_const(0)
        }
    }

    pub fn write_register(&mut self, reg: register::RiscV, val: Source) -> InstructionId {
        self.instructions.push(Instruction::WriteReg {
            dest: reg,
            src: val,
        });

        InstructionId(self.instructions.len() - 1)
    }

    pub fn load_const(&mut self, const_val: u32) -> Id {
        self.instr_with_id(|dest| Instruction::LoadConst {
            dest,
            src: const_val,
        })
    }

    pub fn or(&mut self, src1: Source, src2: Source) -> Id {
        self.binop(BinOp::Or, src1, src2)
    }

    pub fn sll(&mut self, src1: Source, src2: Source) -> Id {
        self.binop(BinOp::Sll, src1, src2)
    }

    pub fn srl(&mut self, src1: Source, src2: Source) -> Id {
        self.binop(BinOp::Srl, src1, src2)
    }

    pub fn sra(&mut self, src1: Source, src2: Source) -> Id {
        self.binop(BinOp::Sra, src1, src2)
    }

    pub fn xor(&mut self, src1: Source, src2: Source) -> Id {
        self.binop(BinOp::Xor, src1, src2)
    }

    pub fn and(&mut self, src1: Source, src2: Source) -> Id {
        self.binop(BinOp::And, src1, src2)
    }

    pub fn add(&mut self, src1: Source, src2: Source) -> Id {
        self.binop(BinOp::Add, src1, src2)
    }

    pub fn sub(&mut self, src1: Source, src2: Source) -> Id {
        self.binop(BinOp::Sub, src1, src2)
    }

    pub fn cmp(&mut self, src1: Source, src2: Source, mode: CmpKind) -> Id {
        self.instr_with_id(|dest| Instruction::Cmp {
            dest,
            src1,
            src2,
            kind: mode,
        })
    }

    pub fn select(&mut self, cond: Source, if_true: Source, if_false: Source) -> Id {
        self.instr_with_id(|dest| Instruction::Select {
            dest,
            if_true,
            if_false,
            cond,
        })
    }

    pub fn fence(&mut self) {
        self.instructions.push(Instruction::Fence)
    }

    pub fn ret_with_code(mut self, addr: Source, code: Source) -> Vec<Instruction> {
        self.instructions.push(Instruction::Ret { addr, code });

        // todo: assert well formedness here.
        self.instructions
    }

    pub fn ret_with_addr(self, addr: Source) -> Vec<Instruction> {
        self.ret_with_code(addr, Source::Val(0))
    }

    pub fn ret(self) -> Vec<Instruction> {
        let pc = self.pc;
        self.ret_with_addr(Source::Id(pc))
    }
}

impl std::ops::Index<InstructionId> for Context {
    type Output = Instruction;

    fn index(&self, idx: InstructionId) -> &Self::Output {
        &self.instructions[idx.0]
    }
}

pub fn lower_non_terminal(ctx: &mut Context, instr: instruction::Info) {
    let instruction::Info {
        instruction,
        start_address: _start_address,
        len,
    } = instr;

    match instruction {
        instruction::Instruction::J(_)
        | instruction::Instruction::Sys(_)
        | instruction::Instruction::B(_)
        | instruction::Instruction::IJump(_) => panic!("Instruction was a terminal"),

        instruction::Instruction::IMem(instruction::IMem {
            opcode,
            imm: _imm,
            rs1: _rs1,
            rd: _rd,
        }) => {
            match opcode {
                // todo: finer grained fences.
                opcode::IMem::FENCE => {
                    ctx.fence();
                    ctx.add_pc(Source::Val(len));
                }
                opcode::IMem::LD(_) => todo!(),
                opcode::IMem::LDU(_) => todo!(),
            }
        }

        instruction::Instruction::I(instruction::I {
            opcode,
            imm,
            rs1,
            rd,
        }) => {
            let imm = imm as i16 as u32;
            let src = ctx.load_register(rs1);

            let res = match opcode {
                opcode::I::ADDI => ctx.add(Source::Id(src), Source::Val(imm)),
                opcode::I::SICond(cmp_mode) => {
                    ctx.cmp(Source::Id(src), Source::Val(imm), cmp_mode.into())
                }
                opcode::I::XORI => ctx.xor(Source::Id(src), Source::Val(imm)),
                opcode::I::ORI => ctx.or(Source::Id(src), Source::Val(imm)),
                opcode::I::ANDI => ctx.and(Source::Id(src), Source::Val(imm)),
                opcode::I::SLLI => ctx.sll(Source::Id(src), Source::Val(imm)),
                opcode::I::SRLI => ctx.srl(Source::Id(src), Source::Val(imm)),
                opcode::I::SRAI => ctx.sra(Source::Id(src), Source::Val(imm)),
            };

            if let Some(rd) = rd {
                ctx.write_register(rd, Source::Id(res));
            }

            ctx.add_pc(Source::Val(len));
        }

        instruction::Instruction::R(instruction::R {
            rs1,
            rs2,
            rd,
            opcode,
        }) => {
            let src1 = ctx.load_register(rs1);
            let src2 = ctx.load_register(rs2);

            let res = match opcode {
                opcode::R::ADD => ctx.add(Source::Id(src1), Source::Id(src2)),
                opcode::R::SUB => ctx.sub(Source::Id(src1), Source::Id(src2)),
                opcode::R::SLL => ctx.sll(Source::Id(src1), Source::Id(src2)),
                opcode::R::SCond(cmp) => ctx.cmp(Source::Id(src1), Source::Id(src2), cmp.into()),
                opcode::R::XOR => ctx.xor(Source::Id(src1), Source::Id(src2)),
                opcode::R::SRL => ctx.srl(Source::Id(src1), Source::Id(src2)),
                opcode::R::SRA => ctx.sra(Source::Id(src1), Source::Id(src2)),
                opcode::R::OR => ctx.or(Source::Id(src1), Source::Id(src2)),
                opcode::R::AND => ctx.and(Source::Id(src1), Source::Id(src2)),
                opcode::R::MUL => todo!(),
                opcode::R::MULH => todo!(),
                opcode::R::MULHSU => todo!(),
                opcode::R::MULHU => todo!(),
                opcode::R::DIV => todo!(),
                opcode::R::DIVU => todo!(),
                opcode::R::REM => todo!(),
                opcode::R::REMU => todo!(),
            };

            if let Some(rd) = rd {
                ctx.write_register(rd, Source::Id(res));
            }

            ctx.add_pc(Source::Val(len));
        }
        instruction::Instruction::S(_) => todo!(),
        instruction::Instruction::U(instruction::U { opcode, imm, rd }) => {
            // ensure that the immediate only uses the upper 20 bits.
            let imm = imm & 0xffff_f000;

            let res = match opcode {
                opcode::U::LUI => ctx.load_const(imm),
                opcode::U::AUIPC => ctx.add(Source::Id(ctx.pc), Source::Val(imm)),
            };

            if let Some(rd) = rd {
                ctx.write_register(rd, Source::Id(res));
            }

            ctx.add_pc(Source::Val(len));
            todo!()
        }
    }
}

pub fn lower_terminal(mut ctx: Context, instr: instruction::Info) -> Vec<self::Instruction> {
    let instruction::Info {
        instruction,
        start_address: _start_address,
        len,
    } = instr;

    match instruction {
        instruction::Instruction::J(instruction::J {
            opcode: opcode::J::JAL,
            rd,
            imm,
        }) => {
            // load pc + len into rd (the link register)
            if let Some(rd) = rd {
                // we specifically need to avoid modifying `pc`, since we need the
                // old value for adding the immediate.
                let next_pc = ctx.add(Source::Id(ctx.pc), Source::Val(len));

                ctx.write_register(rd, Source::Id(next_pc));
            }

            ctx.add_pc(Source::Val(imm));

            ctx.ret()
        }

        instruction::Instruction::IJump(instruction::IJump {
            opcode: opcode::IJump::JALR,
            rd,
            imm,
            rs1,
        }) => {
            let imm = imm as i16 as u32;

            // load pc + len into rd (the link register)
            if let Some(rd) = rd {
                // we specifically need to avoid modifying `pc`, since we need the
                // old value for adding the immediate.
                let next_pc = ctx.add(Source::Id(ctx.pc), Source::Val(len));

                ctx.write_register(rd, Source::Id(next_pc));
            }

            ctx.add_pc(Source::Val(imm));

            // adding 0 is the same as not adding, so don't bother with a match here.
            if let Some(src) = rs1 {
                let src = ctx.read_register(src);
                ctx.add_pc(Source::Id(src));
            }

            ctx.ret()
        }

        instruction::Instruction::Sys(instruction::Sys { opcode }) => {
            use crate::ReturnCode;
            let return_code = match opcode {
                opcode::RSys::EBREAK => ReturnCode::EBreak,
                opcode::RSys::ECALL => ReturnCode::ECall,
            } as u32;

            let pc = ctx.add_pc(Source::Val(len));

            ctx.ret_with_code(Source::Id(pc), Source::Val(return_code))
        }

        instruction::Instruction::B(instruction::B {
            imm,
            rs1,
            rs2,
            cmp_mode,
        }) => {
            let src1 = ctx.load_register(rs1);
            let src2 = ctx.load_register(rs2);

            let cmp = ctx.cmp(Source::Id(src1), Source::Id(src2), cmp_mode.into());

            let current_pc = Source::Id(ctx.pc);

            let continue_pc = ctx.add(current_pc, Source::Val(len));

            let jump_pc = ctx.add(current_pc, Source::Val(imm as i16 as u32));

            ctx.pc = ctx.select(
                Source::Id(cmp),
                Source::Id(jump_pc),
                Source::Id(continue_pc),
            );

            ctx.ret()
        }

        _ => panic!("Instruction wasn't a terminal"),
    }
}

#[cfg(test)]
mod test {
    use super::Context;
    use crate::ssa::lower::lower_non_terminal;
    use crate::{instruction, opcode, register};

    fn cmp_instrs(expected: &[&str], actual: &[super::Instruction]) {
        for (idx, instr) in actual.iter().enumerate() {
            assert_eq!(expected[idx], format!("{}", instr));
            println!("{}", instr);
        }
    }

    #[test]
    fn jal_basic() {
        let ctx = Context::new(0);

        let instrs = super::lower_terminal(
            ctx,
            instruction::Info::new(
                instruction::Instruction::J(instruction::J {
                    imm: 4096,
                    rd: Some(register::RiscV::X4),
                    opcode: opcode::J::JAL,
                }),
                0,
                4,
            ),
        );

        cmp_instrs(
            &[
                "%0 = 0",
                "%1 = add %0, 4",
                "x4 = %1",
                "%2 = add 4096, %0",
                "ret 0, %2",
            ],
            &instrs,
        );
    }

    #[test]
    fn sys_break() {
        let ctx = Context::new(0);

        let instrs = super::lower_terminal(
            ctx,
            instruction::Info::new(
                instruction::Instruction::Sys(instruction::Sys::new(opcode::RSys::EBREAK)),
                0,
                4,
            ),
        );

        cmp_instrs(&["%0 = 0", "%1 = add 4, %0", "ret 1, %1"], &instrs);
    }

    #[test]
    fn addi_nop() {
        let mut ctx = Context::new(0);
        lower_non_terminal(
            &mut ctx,
            instruction::Info::new(
                instruction::Instruction::I(instruction::I::new(0, None, None, opcode::I::ADDI)),
                0,
                4,
            ),
        );

        let instrs = super::lower_terminal(
            ctx,
            instruction::Info::new(
                instruction::Instruction::Sys(instruction::Sys::new(opcode::RSys::EBREAK)),
                4,
                4,
            ),
        );

        cmp_instrs(
            &[
                "%0 = 0",
                "%1 = 0",
                "%2 = add %1, 0",
                "%3 = add 4, %0",
                "%4 = add 4, %3",
                "ret 1, %4",
            ],
            &instrs,
        );
    }

    #[test]
    fn addi_no_dest() {
        let mut ctx = Context::new(0);
        lower_non_terminal(
            &mut ctx,
            instruction::Info::new(
                instruction::Instruction::I(instruction::I::new(
                    50,
                    Some(register::RiscV::X1),
                    None,
                    opcode::I::ADDI,
                )),
                0,
                4,
            ),
        );

        let instrs = super::lower_terminal(
            ctx,
            instruction::Info::new(
                instruction::Instruction::Sys(instruction::Sys::new(opcode::RSys::EBREAK)),
                4,
                4,
            ),
        );

        cmp_instrs(
            &[
                "%0 = 0",
                "%1 = x1",
                "%2 = add %1, 50",
                "%3 = add 4, %0",
                "%4 = add 4, %3",
                "ret 1, %4",
            ],
            &instrs,
        );
    }

    #[test]
    fn addi_no_src() {
        let mut ctx = Context::new(0);
        lower_non_terminal(
            &mut ctx,
            instruction::Info::new(
                instruction::Instruction::I(instruction::I::new(
                    50,
                    None,
                    Some(register::RiscV::X2),
                    opcode::I::ADDI,
                )),
                0,
                4,
            ),
        );

        let instrs = super::lower_terminal(
            ctx,
            instruction::Info::new(
                instruction::Instruction::Sys(instruction::Sys::new(opcode::RSys::EBREAK)),
                4,
                4,
            ),
        );

        cmp_instrs(
            &[
                "%0 = 0",
                "%1 = 0",
                "%2 = add %1, 50",
                "x2 = %2",
                "%3 = add 4, %0",
                "%4 = add 4, %3",
                "ret 1, %4",
            ],
            &instrs,
        );
    }
}
