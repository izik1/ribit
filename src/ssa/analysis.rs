use crate::ssa::{Id, Instruction, Source, StackIndex};
use std::collections::HashMap;
use std::convert::TryFrom;
use std::fmt;

fn lifetime_instruction<F: FnMut(&mut Lifetimes, Source, usize)>(
    instr: &Instruction,
    idx: usize,
    lifetimes: &mut Lifetimes,
    mut update: F,
) {
    match instr {
        Instruction::Fence => {}

        Instruction::BinOp {
            dest, src1, src2, ..
        }
        | Instruction::Cmp {
            dest, src1, src2, ..
        } => {
            update(lifetimes, Source::Id(*dest), idx);
            update(lifetimes, *src1, idx);
            update(lifetimes, *src2, idx);
        }

        Instruction::ReadReg { dest, base, .. } => {
            update(lifetimes, *base, idx);
            update(lifetimes, Source::Id(*dest), idx);
        }

        Instruction::Arg { dest, .. }
        | Instruction::LoadConst { dest, .. }
        | Instruction::ReadStack { dest, .. } => {
            update(lifetimes, Source::Id(*dest), idx);
        }

        Instruction::WriteStack { dest: _, src } => {
            update(lifetimes, Source::Id(*src), idx);
        }

        Instruction::WriteReg { src, base, .. } => {
            update(lifetimes, *base, idx);
            update(lifetimes, *src, idx);
        }

        Instruction::ReadMem {
            dest, src, base, ..
        } => {
            update(lifetimes, *base, idx);
            update(lifetimes, Source::Id(*dest), idx);
            update(lifetimes, *src, idx);
        }

        Instruction::WriteMem {
            addr, src, base, ..
        } => {
            update(lifetimes, *base, idx);
            update(lifetimes, *addr, idx);
            update(lifetimes, *src, idx);
        }

        Instruction::Select {
            dest,
            cond,
            if_true,
            if_false,
        } => {
            update(lifetimes, Source::Id(*dest), idx);
            update(lifetimes, *cond, idx);
            update(lifetimes, *if_true, idx);
            update(lifetimes, *if_false, idx);
        }

        Instruction::Ret { addr, code } => {
            update(lifetimes, *addr, idx);
            update(lifetimes, *code, idx);
        }
    }
}

#[derive(Copy, Clone, Eq, PartialEq)]
pub struct Lifetime {
    pub start: usize,
    pub end: usize,
}

impl Lifetime {
    #[must_use]
    pub fn is_alive_after(self, idx: usize) -> bool {
        idx >= self.start && idx <= self.end
    }

    #[must_use]
    pub fn is_empty(self) -> bool {
        self.start == self.end
    }
}

pub type Lifetimes = HashMap<Id, Lifetime>;

fn update_lifetime(lifetimes: &mut Lifetimes, src: Source, idx: usize) {
    if let Some(id) = src.id() {
        lifetimes
            .entry(id)
            .and_modify(|it| it.end = idx)
            .or_insert(Lifetime {
                start: idx,
                end: idx,
            });
    }
}

#[must_use]
pub fn lifetimes(instrs: &[Instruction]) -> Lifetimes {
    let mut lifetimes = HashMap::with_capacity(instrs.len());

    for (idx, instr) in instrs.iter().enumerate() {
        lifetime_instruction(instr, idx, &mut lifetimes, update_lifetime);
    }

    lifetimes
}

#[must_use]
pub fn surrounding_usages(instrs: &[Instruction], needle: usize) -> Lifetimes {
    let mut lifetimes = HashMap::new();

    for (idx, instr) in instrs.iter().enumerate().take(needle) {
        lifetime_instruction(instr, idx, &mut lifetimes, |lifetimes, src, idx| {
            if let Some(id) = src.id() {
                lifetimes.insert(
                    id,
                    Lifetime {
                        start: idx,
                        end: idx,
                    },
                );
            }
        })
    }

    for (idx, instr) in instrs.iter().enumerate().skip(needle) {
        lifetime_instruction(instr, idx, &mut lifetimes, |lifetimes, src, idx| {
            if let Some(id) = src.id() {
                lifetimes.entry(id).and_modify(|it| {
                    if it.is_empty() {
                        it.end = idx
                    }
                });
            }
        })
    }

    lifetimes
}

// todo: replace this with a Vec<...> `rather` than the hashmap,
//  this is because StackIndexes should always be _used_ linearly.
#[must_use]
pub fn stack_lifetimes(graph: &[Instruction]) -> HashMap<StackIndex, Vec<(usize, usize)>> {
    let mut lifetimes = HashMap::new();
    for (idx, instr) in graph.iter().enumerate() {
        match instr {
            Instruction::WriteStack { dest, src: _ } => lifetimes
                .entry(*dest)
                .or_insert_with(Vec::new)
                .push((idx, idx)),
            Instruction::ReadStack { dest: _, src } => {
                lifetimes.get_mut(src).unwrap().last_mut().unwrap().1 = idx;
            }
            _ => {}
        }
    }

    lifetimes
}

/// Get the highest used `StackIndex` for a given graph.
#[must_use]
pub fn max_stack(graph: &[Instruction]) -> Option<StackIndex> {
    let mut max: Option<StackIndex> = None;

    for instr in graph {
        match instr {
            Instruction::WriteStack {
                dest: stack_idx,
                src: _,
            }
            | Instruction::ReadStack {
                dest: _,
                src: stack_idx,
            } => {
                max = std::cmp::max(max, Some(*stack_idx));
            }

            _ => {}
        }
    }

    max
}

#[must_use]
pub fn min_stack(
    lifetime: Lifetime,
    stack_lts: &HashMap<StackIndex, Vec<(usize, usize)>>,
) -> StackIndex {
    stack_lts
        .iter()
        .filter_map(|(idx, stack_lts)| {
            (stack_lts
                .iter()
                .all(|stack_lt| stack_lt.1 < lifetime.start || stack_lt.0 > lifetime.end))
            .then_some(*idx)
        })
        .min()
        .unwrap_or_else(|| {
            StackIndex(u8::try_from(stack_lts.len()).expect("Ran out of stack indexes"))
        })
}

pub struct ShowLifetimes<'a, 'b> {
    lifetimes: &'a Lifetimes,
    instrs: &'b [Instruction],
}

impl<'a, 'b> ShowLifetimes<'a, 'b> {
    pub fn new(lifetimes: &'a Lifetimes, instrs: &'b [Instruction]) -> Self {
        Self { lifetimes, instrs }
    }
}

fn show_lifetime(f: &mut fmt::Formatter, lifetime: &Lifetime, idx: usize) -> fmt::Result {
    let width = (lifetime.end - lifetime.start) / 10 + 1;
    if lifetime.is_alive_after(idx) {
        write!(f, "{:1$}", lifetime.end - idx, width)
    } else {
        f.write_str(&"-".repeat(width))
    }
}

impl fmt::Display for ShowLifetimes<'_, '_> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        for (idx, instr) in self.instrs.iter().enumerate() {
            let mut lifetimes: Vec<Lifetime> = self.lifetimes.values().cloned().collect();
            lifetimes.sort_by(|a, b| a.start.cmp(&b.start));
            let mut lifetimes = lifetimes.iter();

            write!(f, "[")?;

            if let Some(lifetime) = lifetimes.next() {
                show_lifetime(f, lifetime, idx)?;
            }

            for lifetime in lifetimes {
                f.write_str(", ")?;
                show_lifetime(f, lifetime, idx)?;
            }

            f.write_str("] ")?;

            writeln!(f, "{}", instr)?;
        }

        Ok(())
    }
}

#[cfg(test)]
mod test {
    use super::ShowLifetimes;
    use crate::ssa::opt;

    use insta::assert_display_snapshot;

    #[test]
    fn max_lifetimes() {
        let (mut instrs, _) = crate::ssa::max_fn();

        opt::fold_and_prop_consts(&mut instrs);
        let instrs = opt::dead_instruction_elimination(&instrs);
        let instrs = opt::register_writeback_shrinking(&instrs);

        let lifetimes = super::lifetimes(&instrs);

        assert_display_snapshot!(ShowLifetimes::new(&lifetimes, &instrs));
    }
}
