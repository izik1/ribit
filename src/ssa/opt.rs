use crate::ssa::{eval, Id, Instruction, Source};
use std::collections::HashMap;

fn const_id_lookup(consts: &HashMap<Id, u32>, src: Source) -> Option<u32> {
    match src {
        Source::Id(id) => Some(id),
        Source::Val(_) => None,
    }
    .and_then(|it| consts.get(&it).copied())
}

fn const_prop(consts: &HashMap<Id, u32>, src: &mut Source) -> Option<u32> {
    if let Some(v) = const_id_lookup(consts, *src) {
        *src = Source::Val(v);
        Some(v)
    } else if let Source::Val(v) = src {
        Some(*v)
    } else {
        None
    }
}

pub fn fold_and_prop_consts(graph: &mut [Instruction]) {
    let mut consts = HashMap::new();
    for instruction in graph.iter_mut() {
        let (dest, val) = match instruction {
            Instruction::LoadConst { dest, src } => (*dest, *src),

            Instruction::Fence | Instruction::ReadReg { .. } => continue,

            Instruction::WriteReg { src, .. }
            | Instruction::ReadMem { src, .. }
            | Instruction::WriteMem { src, .. } => {
                const_prop(&consts, src);

                continue;
            }

            Instruction::BinOp {
                dest,
                src1,
                src2,
                op,
            } => {
                let src1 = const_prop(&consts, src1);
                let src2 = const_prop(&consts, src2);

                let res = match (src1, src2) {
                    (Some(src1), Some(src2)) => eval::binop(src1, src2, *op),
                    _ => continue,
                };

                (*dest, res)
            }

            Instruction::Cmp {
                dest,
                src1,
                src2,
                kind,
            } => {
                let src1 = const_prop(&consts, src1);
                let src2 = const_prop(&consts, src2);

                // note: this explicitly doesn't simplify things like
                // `cmp eq %0, %0`, that's for a different pass
                // todo: write pass for the above.

                let res = match (src1, src2) {
                    (Some(src1), Some(src2)) => eval::cmp(src1, src2, *kind),
                    _ => continue,
                };

                (*dest, res)
            }

            Instruction::Select {
                dest,
                cond,
                if_true,
                if_false,
            } => {
                let cond = const_prop(&consts, cond);
                let if_true = const_prop(&consts, if_true);
                let if_false = const_prop(&consts, if_false);

                if let Some(res) = eval::try_select(cond, if_true, if_false) {
                    (*dest, res)
                } else {
                    continue;
                }
            }

            Instruction::Ret { addr, code } => {
                const_prop(&consts, addr);
                const_prop(&consts, code);

                continue;
            }
        };

        *instruction = Instruction::LoadConst { dest, src: val };

        consts.insert(dest, val);
    }
}

fn mark_live(live_instructions: &mut [bool; 0x1_0000], src: Source) {
    if let Source::Id(id) = src {
        live_instructions[id.0 as usize] = true;
    }
}

#[must_use]
pub fn dead_instruction_elimination(graph: &[Instruction]) -> Vec<Instruction> {
    let mut live_ids = [false; 0x1_0000];
    let mut live_instruction_count = 0;

    for instruction in graph.iter().rev() {
        match instruction {
            Instruction::Ret { addr, code } => {
                mark_live(&mut live_ids, *addr);
                mark_live(&mut live_ids, *code);

                live_instruction_count += 1;
            }

            Instruction::Fence => live_instruction_count += 1,
            Instruction::ReadReg { dest, .. } | Instruction::LoadConst { dest, .. } => {
                if live_ids[dest.0 as usize] {
                    live_instruction_count += 1;
                }
            }

            Instruction::ReadMem { src, .. } => {
                // hack: I'm not sure when it's safe to remove unused memory reads,
                // so assume we never can.
                live_instruction_count += 1;
                mark_live(&mut live_ids, *src);
            }

            Instruction::WriteReg { src, .. } => {
                live_instruction_count += 1;
                mark_live(&mut live_ids, *src);
            }

            Instruction::WriteMem { addr, src, .. } => {
                live_instruction_count += 1;
                mark_live(&mut live_ids, *addr);
                mark_live(&mut live_ids, *src);
            }

            Instruction::BinOp {
                dest, src1, src2, ..
            }
            | Instruction::Cmp {
                dest, src1, src2, ..
            } => {
                if live_ids[dest.0 as usize] {
                    live_instruction_count += 1;
                    mark_live(&mut live_ids, *src1);
                    mark_live(&mut live_ids, *src2);
                }
            }

            Instruction::Select {
                dest,
                cond,
                if_true,
                if_false,
            } => {
                if live_ids[dest.0 as usize] {
                    live_instruction_count += 1;
                    mark_live(&mut live_ids, *cond);
                    mark_live(&mut live_ids, *if_true);
                    mark_live(&mut live_ids, *if_false);
                }
            }
        }
    }

    let mut instructions = Vec::with_capacity(live_instruction_count);

    for instr in graph.iter().filter(|it| match it.id() {
        Some(id) => live_ids[id.0 as usize],
        None => true,
    }) {
        instructions.push(instr.clone());
    }

    instructions
}

/// Moves register writes to be as close to their computations as possible.
pub fn register_writeback_shrinking(graph: &[Instruction]) -> Vec<Instruction> {
    let mut stores = vec![];
    let mut instrs = Vec::with_capacity(graph.len());

    for instr in graph {
        match instr {
            Instruction::WriteReg {
                dest,
                src: Source::Id(id),
            } => {
                stores.push((*dest, *id));
            }

            _ => instrs.push(instr.clone()),
        }
    }

    for idx in (0..instrs.len()).rev() {
        if let Some(store_idx) = instrs[idx]
            .id()
            .and_then(|id| stores.iter().copied().position(|(_, src)| src == id))
        {
            let store = stores.remove(store_idx);
            // insert _after_ the instruction we just looked at
            instrs.insert(
                idx + 1,
                Instruction::WriteReg {
                    dest: store.0,
                    src: Source::Id(store.1),
                },
            );
        }
    }

    instrs
}

#[cfg(test)]
mod test {
    use crate::ssa::lower;
    use crate::DisplayDeferSlice;
    use crate::{instruction, opcode, register};

    use insta::assert_display_snapshot;

    #[test]
    fn jal_basic_const_prop() {
        let ctx = lower::Context::new(0);

        let mut instrs = lower::terminal(
            ctx,
            instruction::Instruction::J(instruction::J {
                imm: 4096,
                rd: Some(register::RiscV::X4),
                opcode: opcode::J::JAL,
            }),
            4,
        );

        super::fold_and_prop_consts(&mut instrs);

        assert_display_snapshot!(DisplayDeferSlice(&instrs));
    }

    #[test]
    fn max() {
        let mut instrs = crate::ssa::max_fn();

        super::fold_and_prop_consts(&mut instrs);
        let instrs = super::dead_instruction_elimination(&instrs);
        let instrs = super::register_writeback_shrinking(&instrs);

        assert_display_snapshot!(DisplayDeferSlice(&instrs));
    }

    #[test]
    fn jal_basic_die() {
        let ctx = lower::Context::new(0);

        let mut instrs = lower::terminal(
            ctx,
            instruction::Instruction::J(instruction::J {
                imm: 4096,
                rd: Some(register::RiscV::X4),
                opcode: opcode::J::JAL,
            }),
            4,
        );

        super::fold_and_prop_consts(&mut instrs);
        let instrs = super::dead_instruction_elimination(&instrs);

        assert_display_snapshot!(DisplayDeferSlice(&instrs));
    }
}
