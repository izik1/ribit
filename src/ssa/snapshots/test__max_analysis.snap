---
source: src/ssa/analysis.rs
expression: "ShowLifetimes::new(&lifetimes, &instrs)"
---
[7, -, -, -, -, -, -, -] %0 = x10
[6, 1, -, -, -, -, -, -] %1 = x11
[5, 0, 3, -, -, -, -, -] %2 = add %0, %1
[4, -, 2, 2, -, -, -, -] %3 = srl %2, 31
[3, -, 1, 1, -, -, -, -] x12 = %3
[2, -, 0, 0, 2, -, -, -] %4 = and %2, %3
[1, -, -, -, 1, -, -, -] x11 = %4
[0, -, -, -, 0, 1, -, -] %5 = add %0, %4
[-, -, -, -, -, 0, -, -] x10 = %5
[-, -, -, -, -, -, 1, -] %6 = x1
[-, -, -, -, -, -, 0, 1] %7 = add %6, 1040
[-, -, -, -, -, -, -, 0] ret 0, %7

